File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/external/parser/ContextReversoHtmlParser.java
package org.bubbasmith.vocabtivator.external.parser;

import org.bubbasmith.vocabtivator.model.Sentence;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ContextReversoHtmlParser implements HtmlParser {

    @Override
    public List<Sentence> parse(String html) {
        return Jsoup.parse(html).select(".example>.src>.text")
                .stream().map(this::extractSentence).collect(Collectors.toList());
    }

    private Sentence extractSentence(Element element) {
        StringBuilder text = new StringBuilder();
        List<String> placeholders = new ArrayList<>();

        for (Node childNode : element.childNodes()) {
            if (childNode instanceof TextNode) {
                text.append(((TextNode) childNode).text());
            } else if (childNode instanceof Element) {
                if ("em".equals(((Element) childNode).tagName())) {
                    text.append(String.format(PLACEHOLDER_TEMPLATE, placeholders.size()));
                    placeholders.add(((Element) childNode).text());
                }
            }
        }

        return new Sentence()
                .setText(text.toString())
                .setPlaceholders(placeholders);
    }
}

File: vocabtivator-sentence/src/main/java/org/bubbasmith/vocabtivator/integration/parser/ContextReversoHtmlParser.kt
package org.bubbasmith.vocabtivator.integration.parser

import org.bubbasmith.vocabtivator.model.Sentence
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.nodes.TextNode
import org.springframework.stereotype.Service
import java.util.stream.Collectors

@Service
class ContextReversoHtmlParser : HtmlParser {
    override fun parse(html: String): List<Sentence> {
        return Jsoup.parse(html).select(".example>.src>.text")
                .stream().map { element: Element -> extractSentence(element) }.collect(Collectors.toList())
    }

    private fun extractSentence(element: Element): Sentence {
        val text = StringBuilder()
        val placeholders: MutableList<String> = ArrayList()
        for (childNode in element.childNodes()) {
            if (childNode is TextNode) {
                text.append(childNode.text())
            } else if (childNode is Element) {
                if ("em" == childNode.tagName()) {
                    text.append(String.format(HtmlParser.PLACEHOLDER_TEMPLATE, placeholders.size))
                    placeholders.add(childNode.text())
                }
            }
        }
        return Sentence(text.toString(), placeholders)
    }
}